// windicss-webpack-plugin/dist/shared/windicss-webpack-plugin.c33fb8ee.mjs
import _debug from 'debug';

const NAME = "windicss-webpack-plugin";
const MODULE_ID = "windi.css";
const MODULE_ID_VIRTUAL_TEST = /virtual:windi-?(.*?)\.css/;
const MODULE_ID_VIRTUAL_PREFIX = "virtual:windi";
const MODULE_ID_VIRTUAL_MODULES = [
  `${MODULE_ID_VIRTUAL_PREFIX}.css`,
  `${MODULE_ID_VIRTUAL_PREFIX}-base.css`,
  `${MODULE_ID_VIRTUAL_PREFIX}-utilities.css`,
  `${MODULE_ID_VIRTUAL_PREFIX}-components.css`
];
const HAS_DIRECTIVE_TEST = /@(apply|variants|screen|layer)\s/;
const HAS_THEME_FUNCTION_TEST = /theme\(.*?\)/;
const DEVTOOLS_MODULE_ID = "windi-devtools";
const DEVTOOLS_VIRTUAL_MODULE_ID = "virtual:windi-devtools";
const DEVTOOLS_VIRTUAL_MODULE = "virtual:windi-devtools.js";
const DEVTOOLS_POST_PATH = "/@windicss-devtools-update";
const DEFAULT_SERVER_HOST = "127.0.0.1";
const DEFAULT_SERVER_PORT = 8888;

const debug = {
  plugin: _debug(`${NAME}:plugin`),
  loader: _debug(`${NAME}:loader`)
};

const cssRequiresTransform = (source) => {
  return HAS_DIRECTIVE_TEST.test(source) || HAS_THEME_FUNCTION_TEST.test(source);
};
const isJsx = (source) => {
  return /{`(.*)`}/gms.test(source);
};
const transformCSS = (service, source, resource) => {
  if (!source || source.length <= 0)
    return source;
  if (!cssRequiresTransform(source))
    return source;
  let output = source;
  try {
    output = service.transformCSS(source, resource, { globaliseKeyframes: true });
    if (!output || output.length <= 0) {
      debug.loader(`[WindiCSS] Invalid response from windi core transforming resource: ${resource}.`);
      return source;
    }
    debug.loader("Transformed CSS", resource);
  } catch (e) {
    debug.loader(`[WindiCSS] Exception when transforming CSS for resource: ${resource}.`, e);
    return source;
  }
  return output;
};
const def = (val, def2) => {
  if (val)
    return val;
  return def2;
};
function getChangedModuleNames(utils) {
  if (utils.hasPending)
    utils.buildPendingStyles();
  const moduleNames = [
    `${MODULE_ID_VIRTUAL_PREFIX}.css`
  ];
  Object.entries(utils.layersMeta).forEach(([name, meta]) => {
    if (meta.cssCache == null)
      moduleNames.push(`${MODULE_ID_VIRTUAL_PREFIX}-${name}.css`);
  });
  return moduleNames;
}
const isDev = () => process.env.NODE_ENV === "development";
const isWebCompilerTarget = (target) => {
  let isWeb = true;
  if (typeof target === "string") {
    isWeb = !target.includes("node");
  } else if (Array.isArray(target)) {
    target.forEach((str) => {
      if (str.includes("node"))
        isWeb = false;
    });
  } else {
    isWeb = false;
  }
  return isWeb;
};

export { DEFAULT_SERVER_HOST as D, MODULE_ID as M, NAME as N, DEFAULT_SERVER_PORT as a, DEVTOOLS_POST_PATH as b, DEVTOOLS_VIRTUAL_MODULE as c, MODULE_ID_VIRTUAL_MODULES as d, DEVTOOLS_MODULE_ID as e, DEVTOOLS_VIRTUAL_MODULE_ID as f, getChangedModuleNames as g, debug as h, isDev as i, MODULE_ID_VIRTUAL_TEST as j, def as k, isWebCompilerTarget as l, isJsx as m, transformCSS as t };
